internal/storage/sql_provider.go

```diff
@@ -36,50 +43,98 @@ type SQLProvider struct {
 
 	sqlInsertAuthenticationLog     string
 	sqlGetLatestAuthenticationLogs string
+
+	sqlGetExistingTables string
+	sqlCheckTableExists  string
+
+	sqlConfigSetValue string
+	sqlConfigGetValue string
 }
 
+//nolint:gocyclo // TODO: See if this can be simplified.
 func (p *SQLProvider) initialize(db *sql.DB) error {
 	p.db = db
 
-	_, err := db.Exec(p.sqlCreateUserPreferencesTable)
-	if err != nil {
-		return fmt.Errorf("Unable to create table %s: %v", preferencesTableName, err)
-	}
+	p.log = logging.Logger()
 
-	_, err = db.Exec(p.sqlCreateIdentityVerificationTokensTable)
+	rows, err := db.Query(p.sqlGetExistingTables)
 	if err != nil {
-		return fmt.Errorf("Unable to create table %s: %v", identityVerificationTokensTableName, err)
+		return fmt.Errorf("Unable to initialize database: %v", err)
 	}
+	defer rows.Close()
 
-	_, err = db.Exec(p.sqlCreateTOTPSecretsTable)
-	if err != nil {
-		return fmt.Errorf("Unable to create table %s: %v", totpSecretsTableName, err)
-	}
+	var tables []string
 
-	// keyHandle and publicKey are stored in base64 format
-	_, err = db.Exec(p.sqlCreateU2FDeviceHandlesTable)
-	if err != nil {
-		return fmt.Errorf("Unable to create table %s: %v", u2fDeviceHandlesTableName, err)
+	var table string
+
+	for rows.Next() {
+		err := rows.Scan(&table)
+		if err != nil {
+			return fmt.Errorf("Unable to initialize database: %v", err)
+		}
+
+		tables = append(tables, table)
 	}
 
-	_, err = db.Exec(p.sqlCreateAuthenticationLogsTable)
-	if err != nil {
-		return fmt.Errorf("Unable to create table %s: %v", authenticationLogsTableName, err)
+	p.log.Debug("Storage schema is being checked to verify it is up to date")
+
+	version := 0
+
+	if utils.IsStringInSlice(configTableName, tables) {
+		rows, err := p.db.Query(p.sqlConfigGetValue, "schema", "version")
+		if err != nil {
+			return err
+		}
+
+		for rows.Next() {
+			err := rows.Scan(&version)
+			if err != nil {
+				return err
+			}
+		}
 	}
 
-	// Create an index on (username, time) because this couple is highly used by the regulation module
-	// to check whether a user is banned.
-	if p.sqlCreateAuthenticationLogsUserTimeIndex != "" {
-		_, err = db.Exec(p.sqlCreateAuthenticationLogsUserTimeIndex)
+	p.log.Debugf("Storage schema is v%d, latest is v%d", version, storageSchemaCurrentVersion)
+
+	if version < storageSchemaCurrentVersion {
+		tx, err := p.db.Begin()
 		if err != nil {
-			return fmt.Errorf("Unable to create table %s: %v", authenticationLogsTableName, err)
+			return err
+		}
+
+		switch version {
+		case 0:
+			err := p.upgradeSchemaVersionTo001(tx, tables)
+			if err != nil {
+				_ = tx.Rollback()
```

Comment:
I think we can log the error to be able to know what happens if the statement fails.